auth/service.go

```diff
@@ -718,31 +718,63 @@ func (svc service) AssignUsers(ctx context.Context, token, id string, userIds []
 }
 
 func (svc service) UnassignUsers(ctx context.Context, token, id string, userIds []string, relation string) error {
-	if err := svc.Authorize(ctx, PolicyReq{
+	pr := PolicyReq{
 		Subject:     token,
 		SubjectType: UserType,
 		SubjectKind: TokenKind,
 		Object:      id,
 		ObjectType:  DomainType,
 		Permission:  SharePermission,
-	}); err != nil {
+	}
+	if err := svc.Authorize(ctx, pr); err != nil {
 		return err
 	}
 
-	if err := svc.Authorize(ctx, PolicyReq{
-		Subject:     token,
-		SubjectType: UserType,
-		SubjectKind: TokenKind,
-		Object:      id,
-		ObjectType:  DomainType,
-		Permission:  SwitchToPermission(relation),
-	}); err != nil {
-		return err
+	if relation != "" {
+		pr.Permission = SwitchToPermission(relation)
+		if err := svc.Authorize(ctx, pr); err != nil {
+			return err
+		}
+
+		if err := svc.removeDomainPolicies(ctx, id, relation, userIds...); err != nil {
+			return errors.Wrap(errRemovePolicies, err)
+		}
+		return nil
+	}
+	pr.Permission = AdminPermission
+	if err := svc.Authorize(ctx, pr); err != nil {
+		// User is not admin.
+		var ids []string
+		for _, userID := range userIds {
+			if err := svc.Authorize(ctx, PolicyReq{
+				Subject:     userID,
+				SubjectType: UserType,
+				SubjectKind: UsersKind,
+				Permission:  AdminPermission,
+				Object:      id,
+				ObjectType:  DomainType,
+			}); err != nil {
+				// Append all non-admins to ids.
+				ids = append(ids, userID)
+			}
+		}
+
+		for _, rel := range []string{MemberRelation, ViewerRelation, EditorRelation} {
+			// Remove only non-admins.
+			if err := svc.removeDomainPolicies(ctx, id, rel, ids...); err != nil {
+				return errors.Wrap(errRemovePolicies, err)
+			}
+		}
```

Comment:
You can replace this whole section, since you have the loop in the block below:
```suggestion
	userIds = ids
```